In [542]:
import pandas as  pd
import numpy as np
import matplotlib.pyplot as plt
import sklearn

In [543]:
data_set=pd.read_csv("/content/Dataset.csv")
data_set

,Restaurant ID,Restaurant Name,Country Code,City,Address,Locality,Locality Verbose,Longitude,Latitude,Cuisines,...,Currency,Has Table booking,Has Online delivery,Is delivering now,Switch to order menu,Price range,Aggregate rating,Rating color,Rating text,Votes
0,6317637,Le Petit Souffle,162,Makati City,"Third Floor, Century City Mall, Kalayaan Avenu...","Century City Mall, Poblacion, Makati City","Century City Mall, Poblacion, Makati City, Mak...",121.027535,14.565443,"French, Japanese, Desserts",...,Botswana Pula(P),Yes,No,No,No,3,4.8,Dark Green,Excellent,314
1,6304287,Izakaya Kikufuji,162,Makati City,"Little Tokyo, 2277 Chino Roces Avenue, Legaspi...","Little Tokyo, Legaspi Village, Makati City","Little Tokyo, Legaspi Village, Makati City, Ma...",121.014101,14.553708,Japanese,...,Botswana Pula(P),Yes,No,No,No,3,4.5,Dark Green,Excellent,591
2,6300002,Heat - Edsa Shangri-La,162,Mandaluyong City,"Edsa Shangri-La, 1 Garden Way, Ortigas, Mandal...","Edsa Shangri-La, Ortigas, Mandaluyong City","Edsa Shangri-La, Ortigas, Mandaluyong City, Ma...",121.056831,14.581404,"Seafood, Asian, Filipino, Indian",...,Botswana Pula(P),Yes,No,No,No,4,4.4,Green,Very Good,270
3,6318506,Ooma,162,Mandaluyong City,"Third Floor, Mega Fashion Hall, SM Megamall, O...","SM Megamall, Ortigas, Mandaluyong City","SM Megamall, Ortigas, Mandaluyong City, Mandal...",121.056475,14.585318,"Japanese, Sushi",...,Botswana Pula(P),No,No,No,No,4,4.9,Dark Green,Excellent,365
4,6314302,Sambo Kojin,162,Mandaluyong City,"Third Floor, Mega Atrium, SM Megamall, Ortigas...","SM Megamall, Ortigas, Mandaluyong City","SM Megamall, Ortigas, Mandaluyong City, Mandal...",121.057508,14.584450,"Japanese, Korean",...,Botswana Pula(P),Yes,No,No,No,4,4.8,Dark Green,Excellent,229
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9546,5915730,Naml۱ Gurme,208,��stanbul,"Kemanke�� Karamustafa Pa��a Mahallesi, R۱ht۱m ...",Karak�_y,"Karak�_y, ��stanbul",28.977392,41.022793,Turkish,...,Turkish Lira(TL),No,No,No,No,3,4.1,Green,Very Good,788
9547,5908749,Ceviz A��ac۱,208,��stanbul,"Ko��uyolu Mahallesi, Muhittin ��st�_nda�� Cadd...",Ko��uyolu,"Ko��uyolu, ��stanbul",29.041297,41.009847,"World Cuisine, Patisserie, Cafe",...,Turkish Lira(TL),No,No,No,No,3,4.2,Green,Very Good,1034
9548,5915807,Huqqa,208,��stanbul,"Kuru�_e��me Mahallesi, Muallim Naci Caddesi, N...",Kuru�_e��me,"Kuru�_e��me, ��stanbul",29.034640,41.055817,"Italian, World Cuisine",...,Turkish Lira(TL),No,No,No,No,4,3.7,Yellow,Good,661
9549,5916112,A���k Kahve,208,��stanbul,"Kuru�_e��me Mahallesi, Muallim Naci Caddesi, N...",Kuru�_e��me,"Kuru�_e��me, ��stanbul",29.036019,41.057979,Restaurant Cafe,...,Turkish Lira(TL),No,No,No,No,4,4.0,Green,Very Good,901


droping un used columns

In [544]:
data_set= data_set[['Cuisines',
         'Average Cost for two',
         'Aggregate rating',
         'Votes',
         'Price range',
         'Has Online delivery',
         'Has Table booking']]

In [545]:
data_set['Cuisines'] = data_set['Cuisines'].fillna('').apply(lambda x: x.split(',')[0])

/tmp/ipykernel_570/4112434792.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_set['Cuisines'] = data_set['Cuisines'].fillna('').apply(lambda x: x.split(',')[0])


In [546]:
top_cuisines = data_set['Cuisines'].value_counts().head(5).index
data_set = data_set[data_set['Cuisines'].isin(top_cuisines)]

In [547]:
data_set["Average Cost for one person"]=data_set['Average Cost for two']//2
data_set

,Cuisines,Average Cost for two,Aggregate rating,Votes,Price range,Has Online delivery,Has Table booking,Average Cost for one person
5,Chinese,1000,4.4,336,3,No,No,500
15,Cafe,800,3.6,29,3,No,No,400
18,Cafe,600,4.5,118,2,No,No,300
19,Cafe,1500,4.4,392,4,No,Yes,750
22,Fast Food,55,3.0,6,2,No,No,27
...,...,...,...,...,...,...,...,...
9528,Cafe,70,4.2,79,3,No,No,35
9531,Cafe,80,3.8,123,3,No,No,40
9538,Cafe,30,4.9,1042,2,No,No,15
9543,Cafe,35,4.5,761,2,No,No,17


checking  null values

In [548]:
data_set.isnull().sum()

,0
Cuisines,0
Average Cost for two,0
Aggregate rating,0
Votes,0
Price range,0
Has Online delivery,0
Has Table booking,0
Average Cost for one person,0


In [549]:
from sklearn.preprocessing import LabelEncoder,OneHotEncoder
label=LabelEncoder()
data_set['Cuisines']=label.fit_transform(data_set['Cuisines'])


In [550]:
data_set

,Cuisines,Average Cost for two,Aggregate rating,Votes,Price range,Has Online delivery,Has Table booking,Average Cost for one person
5,2,1000,4.4,336,3,No,No,500
15,1,800,3.6,29,3,No,No,400
18,1,600,4.5,118,2,No,No,300
19,1,1500,4.4,392,4,No,Yes,750
22,3,55,3.0,6,2,No,No,27
...,...,...,...,...,...,...,...,...
9528,1,70,4.2,79,3,No,No,35
9531,1,80,3.8,123,3,No,No,40
9538,1,30,4.9,1042,2,No,No,15
9543,1,35,4.5,761,2,No,No,17


In [551]:
data_set['Has Online delivery'] = data_set['Has Online delivery'].map({'Yes':1, 'No':0})
data_set['Has Table booking'] = data_set['Has Table booking'].map({'Yes':1, 'No':0})


In [552]:
data_set

,Cuisines,Average Cost for two,Aggregate rating,Votes,Price range,Has Online delivery,Has Table booking,Average Cost for one person
5,2,1000,4.4,336,3,0,0,500
15,1,800,3.6,29,3,0,0,400
18,1,600,4.5,118,2,0,0,300
19,1,1500,4.4,392,4,0,1,750
22,3,55,3.0,6,2,0,0,27
...,...,...,...,...,...,...,...,...
9528,1,70,4.2,79,3,0,0,35
9531,1,80,3.8,123,3,0,0,40
9538,1,30,4.9,1042,2,0,0,15
9543,1,35,4.5,761,2,0,0,17


In [553]:
data_set.columns


Index(['Cuisines', 'Average Cost for two', 'Aggregate rating', 'Votes',
       'Price range', 'Has Online delivery', 'Has Table booking',
       'Average Cost for one person'],
      dtype='object')

In [554]:
boolean_cols=data_set.select_dtypes(include="bool").columns
for i in boolean_cols:
  data_set[i]=data_set[i].astype("int")

In [555]:
data_set

,Cuisines,Average Cost for two,Aggregate rating,Votes,Price range,Has Online delivery,Has Table booking,Average Cost for one person
5,2,1000,4.4,336,3,0,0,500
15,1,800,3.6,29,3,0,0,400
18,1,600,4.5,118,2,0,0,300
19,1,1500,4.4,392,4,0,1,750
22,3,55,3.0,6,2,0,0,27
...,...,...,...,...,...,...,...,...
9528,1,70,4.2,79,3,0,0,35
9531,1,80,3.8,123,3,0,0,40
9538,1,30,4.9,1042,2,0,0,15
9543,1,35,4.5,761,2,0,0,17


In [556]:
data_set.info()

<class 'pandas.core.frame.DataFrame'>
Index: 5757 entries, 5 to 9550
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   Cuisines                     5757 non-null   int64  
 1   Average Cost for two         5757 non-null   int64  
 2   Aggregate rating             5757 non-null   float64
 3   Votes                        5757 non-null   int64  
 4   Price range                  5757 non-null   int64  
 5   Has Online delivery          5757 non-null   int64  
 6   Has Table booking            5757 non-null   int64  
 7   Average Cost for one person  5757 non-null   int64  
dtypes: float64(1), int64(7)
memory usage: 404.8 KB


In [557]:
from sklearn.model_selection import train_test_split
y=data_set["Cuisines"]
X=data_set.drop("Cuisines",axis=1,inplace=False)

In [558]:
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.2,random_state=42)



In [559]:
from sklearn.preprocessing import StandardScaler
scaler=StandardScaler()
X_train=scaler.fit_transform(X_train)
X_test=scaler.transform(X_test)

In [560]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC

In [561]:
models = {
    "Logistic Regression": LogisticRegression(max_iter=1000),
    "Decision Tree": DecisionTreeClassifier(),
    "Random Forest": RandomForestClassifier(n_estimators=150, max_depth=10),
    "SVM": SVC(),
    "KNN": KNeighborsClassifier()
}

In [565]:
from sklearn.metrics import accuracy_score, classification_report,confusion_matrix

In [568]:
results = {}
for name, model in models.items():
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)

    acc = accuracy_score(y_test, y_pred)
    results[name] = acc

    print("\n", "="*50)
    print(f"{name}")
    print("Accuracy:", acc)
    print(classification_report(y_test, y_pred))
    print("Confusion Matrix\n",confusion_matrix(y_test,y_pred))


Logistic Regression
Accuracy: 0.5303819444444444
              precision    recall  f1-score   support

           0       0.00      0.00      0.00       118
           1       1.00      0.01      0.02       119
           2       0.00      0.00      0.00       187
           3       0.25      0.01      0.02       119
           4       0.53      1.00      0.69       609

    accuracy                           0.53      1152
   macro avg       0.36      0.20      0.15      1152
weighted avg       0.41      0.53      0.37      1152

Confusion Matrix
 [[  0   0   0   1 117]
 [  0   1   0   0 118]
 [  0   0   0   2 185]
 [  0   0   0   1 118]
 [  0   0   0   0 609]]

Decision Tree
Accuracy: 0.4288194444444444
              precision    recall  f1-score   support

           0       0.28      0.23      0.25       118
           1       0.28      0.29      0.28       119
           2       0.23      0.20      0.22       187
           3       0.15      0.13      0.14       119
           4

/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))



Random Forest
Accuracy: 0.5373263888888888
              precision    recall  f1-score   support

           0       0.23      0.08      0.11       118
           1       0.48      0.29      0.36       119
           2       0.40      0.07      0.13       187
           3       0.26      0.16      0.20       119
           4       0.58      0.89      0.70       609

    accuracy                           0.54      1152
   macro avg       0.39      0.30      0.30      1152
weighted avg       0.47      0.54      0.46      1152

Confusion Matrix
 [[  9   4   5  21  79]
 [  4  34   2   6  73]
 [  2   9  14   9 153]
 [  7   6   3  19  84]
 [ 18  18  11  19 543]]

SVM
Accuracy: 0.5295138888888888
              precision    recall  f1-score   support

           0       0.00      0.00      0.00       118
           1       0.00      0.00      0.00       119
           2       0.00      0.00      0.00       187
           3       0.50      0.01      0.02       119
           4       0.53     

/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


In [564]:

print("\nFinal Accuracy Comparison:")
for model, acc in results.items():
    print(f"{model}: {acc:.4f}")


Final Accuracy Comparison:
Logistic Regression: 0.5304
Decision Tree: 0.4271
Random Forest: 0.5399
SVM: 0.5295
KNN: 0.4549


📊 Model Performance Analysis Across Different Cuisines
🔹 Overview

The objective of this task was to develop a machine learning model to classify restaurants based on their cuisines. Multiple classification algorithms were implemented, including Logistic Regression, Decision Tree, Random Forest, SVM, KNN, and Naive Bayes. Among these, the Random Forest model achieved the highest accuracy of 53.99% and was selected as the final model.

🔹 Performance Across Different Cuisines

The model’s performance varies significantly across different cuisine categories:

High-performing cuisines:
  Frequently occurring cuisines such as North Indian and Chinese were predicted with higher accuracy. This is because these classes had a larger number of samples in the dataset, enabling the model to learn their patterns effectively.

Low-performing cuisines:

 Less frequent cuisines such as Italian, Thai, and other minority classes showed lower prediction accuracy. These cuisines had fewer training samples, making it difficult for the model to generalize.

Misclassification patterns:

 The confusion matrix shows that several cuisines are misclassified as dominant cuisines. Many minority class instances are incorrectly predicted as a major cuisine category, indicating prediction bias

1. Class Imbalance

The dataset contains an uneven distribution of cuisines, where some cuisines appear very frequently while others are rare.

Impact:

Model becomes biased toward majority classes

Minority classes have poor prediction performance

2. Feature Overlap

The features used in the model, such as cost, rating, votes, and price range, are not strong indicators of cuisine type.

Impact:

Different cuisines share similar feature values

Model struggles to differentiate between them

3. Multi-Cuisine Limitation

      Many restaurants offer multiple cuisines, but for simplicity, only the primary cuisine was considered.



Reduced classification accuracy

4. Weak Feature Representation

      Cuisine classification cannot be accurately determined using only numerical features.

      Missing features include:

      Menu items (text data)

🏆 Conclusion

The cuisine classification model demonstrates moderate performance, with Random Forest achieving the best accuracy among all tested models. The model performs well for frequently occurring cuisines but struggles with less represented ones due to class imbalance and overlapping feature values.

Although the accuracy is not very high, the project effectively highlights the challenges of multi-class classification in real-world datasets. Future improvements can include balancing the dataset, incorporating additional features such as textual data, and using advanced models to improve performance.